kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/clients/KafClient.java

```diff
@@ -66,11 +75,67 @@ public void produceMessages(String topicName, String bootstrap, String message,
     }
 
     @Override
-    public String consumeMessages(String topicName, String bootstrap, String messageToCheck, int numOfMessages, Duration timeout) {
+    public List<ConsumerRecord<String, String>> consumeMessages(String topicName, String bootstrap, int numOfMessages, Duration timeout) {
         LOGGER.atInfo().log("Consuming messages using kaf");
         String name = Constants.KAFKA_CONSUMER_CLIENT_LABEL + "-kafka-go";
-        List<String> args = List.of("kaf", "-b", bootstrap, "consume", topicName);
+        List<String> args = List.of("kaf", "-b", bootstrap, "consume", topicName, "--output", "json");
         Job goClientJob = TestClientsJobTemplates.defaultKafkaGoConsumerJob(name, args).build();
-        return KafkaUtils.consumeMessages(topicName, name, deployNamespace, goClientJob, messageToCheck, numOfMessages, timeout);
+        String podName = KafkaUtils.createJob(deployNamespace, name, goClientJob);
+        String log = waitForConsumer(podName, numOfMessages, timeout);
+        LOGGER.atInfo().log(log);
+        List<String> logRecords = List.of(log.split("\n"));
+        return getConsumerRecords(topicName, logRecords);
+    }
+
+    private String waitForConsumer(String podName, int numOfMessages, Duration timeout) {
+        String log;
+        try {
+            log = await().alias("Consumer waiting to receive messages")
+                    .ignoreException(KubernetesClientException.class)
+                    .atMost(timeout)
+                    .until(() -> {
+                        if (kubeClient().getClient().pods().inNamespace(deployNamespace).withName(podName).get() != null) {
+                            return kubeClient().logsInSpecificNamespace(deployNamespace, podName);
+                        }
+                        return null;
+                    }, m -> getNumberOfJsonMessages(m) == numOfMessages);
+        }
+        catch (ConditionTimeoutException e) {
+            log = kubeClient().logsInSpecificNamespace(deployNamespace, podName);
+            LOGGER.atInfo().setMessage("Timeout! Received: {}").addArgument(log).log();
+        }
+        return log;
+    }
+
+    private int getNumberOfJsonMessages(String log) {
+        if (log == null) {
+            return 0;
+        }
+
+        List<String> jsonMessages = new ArrayList<>();
+        String[] logLines = log.split("\n");
+
+        for (String message : logLines) {
+            if (TestUtils.getJsonNode(message) != null) {
+                jsonMessages.add(message);
```

Comment:
it is simpler to increment a counter rather than build a list, call size and throw the list away.